File: code/python/ch11/solutions/warmup_05_sol.py
"""
Solution for Warmup 5: Array Intersection
============================================
Chapter 11: Hashing — The Secret Decoder Ring

APPROACH
--------
Convert both arrays to sets, compute the intersection, and return
the result as a sorted list.

TIME COMPLEXITY:  O(n + m + k log k) where k = size of intersection
SPACE COMPLEXITY: O(n + m) for the two sets
"""


def solve(a: list[int], b: list[int]) -> list[int]:
    """Return sorted list of unique common elements."""
    set_a = set(a)
    set_b = set(b)
    return sorted(set_a & set_b)


# ── Do not change anything below this line ──────────────────────────
if __name__ == "__main__":
    line1 = input().strip()
    line2 = input().strip()
    a = list(map(int, line1.split())) if line1 else []
    b = list(map(int, line2.split())) if line2 else []
    print(solve(a, b))

File: code/cpp/ch11/solutions/warmup_05_sol.cpp
/*
 * Solution for Warmup 5: Intersection of Two Arrays
 * Chapter 11: Hashing — The Secret Decoder Ring
 *
 * APPROACH: Insert first array into an unordered_set, then iterate
 *           second array and collect common elements in a result set.
 *           Sort the result before returning.
 * TIME:  O(n + m + k log k) where k = intersection size
 * SPACE: O(n + m)
 */

#include <algorithm>
#include <iostream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>
using namespace std;

vector<int> solve(vector<int> a, vector<int> b) {
    unordered_set<int> set_a(a.begin(), a.end());
    unordered_set<int> found;
    for (int x : b) {
        if (set_a.count(x)) {
            found.insert(x);
        }
    }
    vector<int> result(found.begin(), found.end());
    sort(result.begin(), result.end());
    return result;
}

// -- Do not change anything below this line --------------------------
int main() {
    int n, m;
    cin >> n;
    vector<int> a(n);
    for (int i = 0; i < n; i++) cin >> a[i];
    cin >> m;
    vector<int> b(m);
    for (int i = 0; i < m; i++) cin >> b[i];
    vector<int> result = solve(a, b);
    for (int x : result) cout << x << " ";
    cout << endl;
    return 0;
}
